src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -385,20 +426,255 @@ namespace QgsWms
     }
     else //unknown format
     {
-      restoreOpacities( bkVectorRenderers, bkRasterRenderers, labelTransparencies, labelBufferTransparencies );
-      clearFeatureSelections( selectedLayerIdList );
       throw QgsBadRequestException( QStringLiteral( "InvalidFormat" ),
                                     QStringLiteral( "Output format '%1' is not supported in the GetPrint request" ).arg( formatString ) );
     }
 
-    restoreOpacities( bkVectorRenderers, bkRasterRenderers, labelTransparencies, labelBufferTransparencies );
-    clearFeatureSelections( selectedLayerIdList );
-    QgsWmsConfigParser::removeHighlightLayers( highlightLayers );
-
     delete c;
     return ba;
   }
 
+  bool QgsRenderer::configureComposition( QgsComposition *c, const QgsMapSettings &mapSettings )
+  {
+    if ( !mWmsParameters.dpi().isEmpty() )
+      c->setPrintResolution( mWmsParameters.dpiAsInt() );
+
+    // Composer items
+    QList<QgsComposerItem * > itemList;
+    c->composerItems( itemList );
+    // Composer legends have to be updated after the other items
+    QList<QgsComposerLegend *> composerLegends;
+
+    // Update composer items
+    QList<QgsComposerItem *>::iterator itemIt = itemList.begin();
+    for ( ; itemIt != itemList.end(); ++itemIt )
+    {
+      // firstly the composer maps
+      if ( ( *itemIt )->type() == QgsComposerItem::ComposerMap )
+      {
+        QgsComposerMap *map = qobject_cast< QgsComposerMap *>( *itemIt );
+        if ( !map )
+          continue;
+
+        QgsWmsParametersComposerMap cMapParams = mWmsParameters.composerMapParameters( map->id() );
+
+        QString mapId = "MAP" + QString::number( map->id() );
+        //map extent is mandatory
+        if ( !cMapParams.mHasExtent )
+        {
+          //remove map from composition if not referenced by the request
+          c->removeItem( map );
+          delete map;
+          continue;
+        }
+
+        // Change CRS of map set to "project CRS" to match requested CRS
+        // (if map has a valid preset crs then we keep this crs and don't use the
+        // requested crs for this map item)
+        if ( mapSettings.destinationCrs().isValid() && !map->presetCrs().isValid() )
+          map->setCrs( mapSettings.destinationCrs() );
+
+        QgsRectangle r( cMapParams.mExtent );
+        if ( mWmsParameters.versionAsNumber() >= QgsProjectVersion( 1, 3, 0 ) && mapSettings.destinationCrs().hasAxisInverted() )
+        {
+          r.invert();
+        }
+        map->setNewExtent( r );
+
+        // scale
+        if ( cMapParams.mScale > 0 )
+        {
+          map->setNewScale( cMapParams.mScale );
+        }
+
+        // rotation
+        if ( cMapParams.mRotation )
+        {
+          map->setMapRotation( cMapParams.mRotation );
+        }
+
+        if ( !map->keepLayerSet() )
+        {
+          if ( !mParameters.contains( mapId + ":LAYERS" ) )
```

Comment:
I think this should be managed thanks to QgsWmsParameters (the ultimate goal may be to not have a `mParameters` attribute).